### PR TITLE
Migration to Terraform provider framework

### LINKED ISSUE #57: Edit notes field
### Summary
Add Notes to the optional schema for the onepassword_item resource
### Use cases

It would be helpful to be able to edit the Notes field in onepassword through the Terraform integration. This would provide an additional feature match to what can be done through OnePassword UI.

Our specific request resolve around adding anote to reference that the item was managed by Terraform. We can do this through a tag but the ideal solution would to expose the notes as a modified attribute.

### LINKED ISSUE #76: For database items hostname is not set because label is server not hostname as the item schema expects
### Your environment

Terraform Provider Version: 1.1.4
Connect Server Version: 1.10.0
OS: Terraform Cloud so probably Ubuntu
Terraform Version: Latest (1.4.2)

## What happened?
Hostname for database category items is not set

## What did you expect to happen?
Hostname for database items should be set

## Steps to reproduce
1. Create a database item in 1password
2. Use the data resource for item to access the item
3. The hostname field will be null instead of the value

## Notes & Logs

The problem stems from onepassword\data_source_onepassword_item.go line 211 to 222 specifically line 220
```
	for _, f := range item.Fields {
		switch f.Purpose {
		case "USERNAME":
			data.Set("username", f.Value)
		case "PASSWORD":
			data.Set("password", f.Value)
		case "NOTES":
			data.Set("note_value", f.Value)
		default:
			if f.Section == nil {
				data.Set(strings.ToLower(f.Label), f.Value)
			}
		}
	}
```

In the case of database items 1connect sends and id for hostname but the label is set to server they probably should both be server as that makes more sense for a database. 

The best work around I could come up with that didn't require a major refactor was to set that section like this. 

```
	for _, f := range item.Fields {
		switch f.Purpose {
		case "USERNAME":
			data.Set("username", f.Value)
		case "PASSWORD":
			data.Set("password", f.Value)
		case "NOTES":
			data.Set("note_value", f.Value)
		default:
			if f.Section == nil {
				if f.Label == "server" {
					data.Set(strings.ToLower(f.ID), f.Value)
				} else {
					data.Set(strings.ToLower(f.Label), f.Value)
				}
			}
		}
	}
```

Basically use ID inste4ad of label for server this gets around the schema validation. Another option would be to solve the naming issue in 1password connect and use hostname or server for both ID and label.

### LINKED ISSUE #149: Support for Secure Note
### Summary
Currently the onepassword provider can only manage items of types (login, password, database), where onepassword supports about 21 types.

This feature request is specifically for the Secure Note type, but it would be great if all type could be supported in the (near) future

We use the Secure Note all the time, but unfortunately this type cannot be created or updated using the 1Password/onepassword provider.

Currently we are leveraging the anasinnyk/onepassword provider for this, but this requires to use the onepassword cli 1.4.0 which is very old.

### Use cases
Larger secure texts
JSON objects
credential files
SSH Keys
etc.

### LINKED ISSUE #155: onepassword_item tags not in alphabetical order have constant diffs
terraform 1.7.4
provider 1.4.2
1passsord CLI 2.25.0

Tags not in alphabetical order cause constant diffs:

```
resource "onepassword_item" "item" {
...
...

  tags = [
    "a",
    "c",
    "b"
  ]
}
```

If I alter the tags to be as follows, then there is not a constant diff.

```
  tags = [
    "a",
    "b",
    "c"
  ]
```

### LINKED ISSUE #157: onepassword_item with login category, cannot remove the password box in 1password unless the password parameter is commented out
terraform 1.7.4
provider 1.4.2
1password CLI 2.25.0

```
resource "onepassword_item" "item" {
...
...
  username = var.myvar != null ? "test : ""
  password = var.myvar != null ? "test : "there is no password"
}
```

I want to REMOVE the password box from 1password and not have it say "there is no password".

But the only way to remove is is to comment out the `password` line, which obviously won't work in my use case.

Setting `there is no password` to null or `""` doesn't remove the password box, in fact it doesnt even update what is currently set for the password.

### PR DESCRIPTION
**Pull request by jillianwilson**
# Migration to Terraform provider framework

## TL;DR

Due to a [bug](https://github.com/hashicorp/terraform-plugin-sdk/issues/201) in the Hashicorp `terraform-plugin-sdk` it is possible for sensitive data pulled from 1Password items to be shown in plaintext when a user runs `terraform plan`. _This only affects the sensitive data pulled from custom sections within 1Password items._

Affected sensitive data pulled from 1Password, _that isn’t marked sensitive in the Terraform plan_, is only shown to the user running `terraform plan` or any third-party terraform providers used to run the plan. **The sensitive data is otherwise not unmasked to any other sources.**

Previously the 1Password Terraform provider was built with the latest version of the [terraform-plugin-sdk](https://github.com/hashicorp/terraform-plugin-sdk) which was impacted by this bug. We have since upgraded our provider to use the [terraform-plugin-framework](https://github.com/hashicorp/terraform-plugin-framework) which no longer has this bug.

## Determining impact

In a Terraform plan, a block can be marked as sensitive so that fields cannot be shown in plaintext. Due to the bug in terraform-plugin-sdk, if nested blocks are marked as sensitive, the property is ignored and sensitive fields can be shown in plaintext.

Have the team responsible for the implementation of the Terraform provider determine if — in their Terraform configuration file — they have any provider that consumes data from an item’s section or if they output it.

This looks something like this:

```tf
resource "checkly_environment_variable" "variable_1" {
      value  = data.onepassword_item.my_login.section[0].field[0].value
}
```

Or like this:

```tf
output "poc" {
    value = jsonencode(data.onepassword_item.my_login.section[0])
}
```

Note that if your configuration is impacted, sensitive data is only viewable in plaintext in _authenticated_ flows. This means that only the users and third-party integrations that are authorized to read those secrets from 1Password can print or read them as plaintext. **Affected sensitive data is not otherwise unmasked to any other sources.**

## What this PR does

The PR migrates the 1Password Terraform provider from using the [terraform-plugin-sdk](https://github.com/hashicorp/terraform-plugin-sdk) to using [terraform-plugin-framework](https://github.com/hashicorp/terraform-plugin-framework). The new framework has a more robust handling of the schema and nested blocks, thereby no longer ignoring the sensitivity of a field’s value from a custom section. The outcome is that now, Terraform will take appropriate action to mask the field and mark it as sensitive. As an example, if a user tries to print an item’s section, Terraform will display an appropriate error.

This fix will be available in the upcoming Terraform provider release v2.0.0 shortly. Please update to that version if you are affected. This update does not contain any breaking changes. In the meantime, if applicable, remove all outputs of Terraform plans and configurations that may contain sensitive data.

### Additional improvements as part of this PR

With the migration to the new [terraform-plugin-framework](https://github.com/hashicorp/terraform-plugin-framework), we've had the chance to improve the provider. Here's a list with the improvements:

- **New**
  - Add support for `Secure Note` items. Resolves #149.
  - Add `note_value` attribute representing a 1Password Item's `notes` field. Resolves #57.
- **Improved**
  - The data handling is more robust, making it less prone to errors and inconsistencies. Resolves #157, #146.
  - CLI and Connect clients now have a more consistent behavior.
  - Fields of type `OTP` are better handled when user provides a custom ID for them. Terraform will throw an error if the custom ID doesn't have the `TOTP_` prefix, which is required for this field type.
  - The values that are generated will only show in the plan to be recomputed when the recipe is changed or the value is explicitly set.
  - When fetching Database items from 1Password, the `server` field (previously known as `hostname`) will populate the Terraform `hostname` attribute. This ensures that the data from new Database items is mapped as expected. Resolves #76.
  - Vault description is now fetched when getting a vault from 1Password by name and the provider was configured to use the CLI client.
- **Fixed**
  - Generated values (using a recipe) are now regenerated when the recipe is changed.
  - Tag ordering mismatch between Terraform state and 1Password no longer causes a change if the tags are the same. The mismatch can be caused by 1Password storing the tags in alphabetical order. Resolves #155.

### Setup

1. Checkout this branch:

   ```sh
   git checkout feat/migrate-to-provider-framework
   ```

2. Build the provider:

   ```sh
   make build
   ```

3. Get your full path to the locally-built provider:

   ```sh
   echo $PwD/dist
   ```

4. Configure your `~/.terraformrc` file to use the local 1Password Terraform provider:

   ```
   provider_installation {

     dev_overrides {
         "1Password/onepassword" = "<path-from-step-3>"

         # Other overrides
     }

     # For all other providers, install them directly from their origin provider
     # registries as normal. If you omit this, Terraform will _only_ use
     # the dev_overrides block, and so no other providers will be available.
     direct {}
   }
   ```

### Test that the bug is fixed

1. In a Terraform file, add the following snippets:

   - Get an item from a vault:

     ```tf
     data "onepassword_vault" "test-vault" {
         name = "<your-vault-name>"
     }

     data "onepassword_item" "test-item" {
         vault = data.onepassword_vault.test-vault.uuid
         title = "<your-item-title>"
     }
     ```

   - Print an item's section to the output:

     ```tf
     output "poc" {
         value = jsonencode(data.onepassword_item.test-item.section[0])
     }
     ```

2. Run a terraform plan:

   ```sh
   terraform plan
   ```

   - [x] It shuold throw the following error:

     ```
     Planning failed. Terraform encountered an error while generating this plan.

     ╷
     │ Error: Output refers to sensitive values
     │
     │   on main.tf line 79:
     │   79: output "poc" {
     │
     │ To reduce the risk of accidentally exporting sensitive data that was
     │ intended to be only internal, Terraform requires that any root module
     │ output containing sensitive data be explicitly marked as sensitive,
     │ to confirm your intent.
     │
     │ If you do intend to export this data, annotate the output value as sensitive
     │ by adding the following argument:
     │     sensitive = true
     ╵
     ```

#### Vault data source

1. Create a Terraform file, add the following snippet:

   ```tf
   terraform {
     required_providers {
       onepassword = {
         source  = "1Password/onepassword"
       }
     }
   }

   provider "onepassword" {}

   data "onepassword_vault" "test-vault" {
       name = "<your-vault-name>"
   }

   output "vault_res" {
       value = jsonencode(data.onepassword_vault.test-vault)
   }
   ```

2. Make a plan for the configuration:

   ```tf
   terraform plan
   ```

   - [x] The `Changes to outputs` section should look like this one:

     ```
     vault_res = jsonencode(
         {
           + description = "This is a description of my vault"
           + id          = "vaults/c2fcrvudd2ptjezfmmrkzxoawq"
           + name        = "test"
           + uuid        = "c2fcrvudd2ptjezfmmrkzxoawq"
         }
     )
     ```

#### Item data source

1. Create a Terraform file, add the following snippet:

   ```tf
   terraform {
     required_providers {
       onepassword = {
         source  = "1Password/onepassword"
       }
     }
   }

   provider "onepassword" {}

   data "onepassword_vault" "test-vault" {
       name = "<your-vault-name>"
   }

   data "onepassword_item" "test-item" {
       vault = data.onepassword_vault.test-vault.uuid
       title = "<your-item-title>"
   }
   ```

2. Make a plan for the configuration:

   ```tf
   terraform plan
   ```

   - [x] It should succeed without any failures.

#### Item resource

1. In a Terraform file, add a complex `onepassword_item` resource. You can use the snippets below as an inspiration.

   <details><summary>Login item example snippet</summary>

   ```tf
   resource "onepassword_item" "demo_login" {
       vault = data.onepassword_vault.test-vault.uuid

       title = "Test TF Resource - Login"
       category = "login"

       username = "MyUserName123"
       password_recipe {
         length = 30
         letters=false
       }

       url = "http://google.com"
   }
   ```

   </details>

   <details><summary>Database item example snippet</summary>

   ```tf
   resource "onepassword_item" "demo_database" {
       vault = data.onepassword_vault.test-vault.uuid

       title = "Demo TF Resource - Database"
       category = "database"

       type = "MySQL"
       hostname = "http://my.fancydomain.com"
       port = "3000"
       database = "badass"
       username = "theboss"
   }
   ```

   </details>

   <details><summary>Secure Note example snippet (new feature)</summary>

   ```tf
   resource "onepassword_item" "test-secure-note" {
       vault = data.onepassword_vault.test-vault.uuid

       title = "Test TF Resource - Secure Note"
       category = "secure_note"

       note_value = <<EOT
       Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec et volutpat lorem. Integer quis felis porttitor, lobortis purus id, sodales nisi. Suspendisse dictum nibh lorem, vel dapibus sapien sodales pulvinar. Donec euismod, dolor vitae fermentum faucibus, elit orci ullamcorper nisi, eget ultrices turpis elit non odio. Integer dictum risus ante, finibus venenatis sem varius bibendum. Sed pellentesque ultrices diam vel viverra. Duis odio mi, malesuada quis luctus a, lacinia quis orci. Vestibulum hendrerit lorem a odio tristique, quis blandit purus sollicitudin.

       Nullam ut hendrerit lacus, at consectetur risus. Vestibulum at diam a arcu fermentum eleifend vel at mi. Cras auctor, massa in blandit rutrum, risus nibh euismod massa, quis porttitor libero mauris id nibh. Sed lobortis ipsum in ipsum dapibus, commodo eleifend velit dignissim. Donec tincidunt turpis mi, vitae venenatis urna laoreet vel. Integer ut scelerisque massa, at varius est. Proin porta quam a dolor euismod, quis iaculis odio dictum. Cras velit neque, vestibulum et consequat at, ultrices quis enim. Nulla sed augue ac odio scelerisque facilisis. Nunc ultricies tristique eros et ullamcorper. Curabitur et est eu nibh viverra feugiat. Pellentesque viverra vitae lacus sed consectetur. Aliquam nec tempor ipsum, quis dapibus quam. Nulla suscipit enim vel porta egestas. Nulla interdum congue erat at fermentum. Quisque viverra tincidunt metus non malesuada.
       EOT
   }
   ```

   </details>

   <details><summary>Complex item example snippet</summary>

   ```tf
   resource "onepassword_item" "demo_complex" {
       vault = data.onepassword_vault.test-vault.uuid

       title = "Test TF Resource - Complex"
       category = "login"

       username = "monseur.user@crazy-IT.com"
       password_recipe {
         length = 64
         digits = false
       }

       url = "https://github.com"

       tags = ["Terraform", "1password-connect", "test"]

       section {
           label = "Some addition info 🛣️"

           field {
               label = "Destination"
               value = "⛰️ Mount Everest"
           }

           field {
               label = "📧 Information"
               type = "EMAIL"
               value = "awesome-location@cabin.nl"
           }

           # This is a new feature added with the migrated framework
           field {
               label = "Mobile"
               type = "PHONE"
               value = "+154242342343128"
           }

           field {
               id = "trip-creation-time"
               label = "Planned since"
               type = "MONTH_YEAR"
               value = "202201"
           }

           field {
               label = "Check-in"
               type = "DATE"
               value = "2024-05-03"
           }

           field {
               label = "Check-out"
               type = "DATE"
               value = "2024-05-10"
           }
       }

       section {
           label = "Other details"

           field {
               label = "Rub rub said..."
               type = "CONCEALED"

               password_recipe {
                 length = 21
                 letters = false
                 symbols = false
               }
           }

           field {
               id = "TOTP_key-to-success"
               label = "Fine, you may leave..."
               type = "OTP"
               value = "otpauth://totp/IT-Tools:demo-user?issuer=IT-Tools&secret=TWODB7HMGDABG3LY&algorithm=SHA1&digits=6&period=30"
           }

           field {
               label = "Protocol"
               type = "URL"
               value = "https://www.robtopgames.com/"
           }
       }

       section {
           label = "🍜 Some last bits"

           field {
               label = "Airline"
               value = "✈️ Qatar"
           }

           field {
               label = "Reach us at"
               type = "PHONE"
               value = "+322134697057"
           }

           field {
               id = "custom-travel-code"
               label = "coupon_code"
               type = "CONCEALED"

               password_recipe {
                 length = 12
                 symbols = false
               }
           }
       }
   }
   ```

   </details>

2. Apply the changes to your configuration to create the 1Password item:

   ```sh
   terraform apply
   ```

3. Check the created item in 1Password

   - [ ] The item should be succesfully created with the appropriate data

4. Make a change to the `onepassword_item` resource. A couple of examples:

   - Change the recipe for the password
   - Add a new section with fields
   - Change the password recipe of a custom field
   - Add a `note_value` to the item (new feature)

5. Apply the changes again and then check the item in 1Password again.

   - [ ] The item should reflect the latest changes.

6. Remove the code for the `onepassword_item` resource and apply the changes again.
   - [ ] The item in 1Password shuold be deleted.

### CODE REVIEW

**Comment by volodymyrZotov** [comment]
✅ Tested it and could confirm that this fixes the issue with printing sensitive data in the sections. 

✅ In addition, I tested other scenarios (create/read/update/delete) items, and it works well after migration. 

⚠️ However, there is an issue with  `secure_note` item that I faced.
To reproduce it, I used the `Secure Note snippet` from the testing steps and when running `terraform apply` I saw the following error in the console:
```
╷
│ Error: Provider returned invalid result object after apply
│ 
│ After the apply operation, the provider still indicated an unknown value for onepassword_item.test-secure-note.password. All values must be known after apply, so this is always a bug in the provider and should be reported in the
│ provider's own repository. Terraform will still save the other known object values in the state.
╵
```
Nevertheless, there is an error in the console, I see that the Secure Note item is created in 1Password.

I approve this PR as it contains an important fix to not print sensitive data in the sections, but let's address the Security Note issue in the following PR.

**Review by volodymyrZotov — APPROVED** [review]

**Comment by volodymyrZotov** [comment]
Filed an issue for the secure note https://github.com/1Password/terraform-provider-onepassword/issues/173
